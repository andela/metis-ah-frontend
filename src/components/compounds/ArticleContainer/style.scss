.article-inner-container {
	padding: 55px 10px;
	flex: 1;
	display: flex;
	flex-direction: column;
	.article-loader {
		display: flex;
		flex:1;
		width: 100%;
    height: 64px;
    justify-content: center;
    align-items: center
	}
	.article-loader:after {
		content: ' ';
		display: block;
		width: 56px;
		height: 56px;
		margin: 1px;
		border-radius: 50%;
		border: 5px solid #13b47a;
		border-color: #13b47a transparent #13b47a transparent;
		animation: article-loader 1.2s linear infinite;
	}

	.article-container-card {
    background-color: white;
		border-radius: 5px;
		display: flex;
		flex-direction: column;
		width: 380px;
		padding: 1%;
		margin: auto;
		align-items: center;
		box-shadow: 0 2px 4px rgba(0,0,0,0.11), 0 2px 4px rgba(0,0,0,0.18);
		.notification-icon {
			font-size: 30px;
			color: #13b47a;
		}
		p {
			font-size: 20px;
			padding-top: 15px;
		}
  }

	@keyframes article-loader {
		0% {
			transform: rotate(0deg);
		}
		100% {
			transform: rotate(360deg);
		}
	}
}

@media screen and (max-width: 425px) {
	.article-inner-container {
		.article-container-card {
			width: 280px;
			.notification-icon {
				font-size: 25px;
			}
			p {
				font-size: 17px;
			}
		}
	}
}

@media screen and (max-width: 768px) {
	.article-inner-container {
		padding: 35px 15px;
	}
}
