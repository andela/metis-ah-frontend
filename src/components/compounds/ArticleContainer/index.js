import React, { Component } from 'react';
import { withRouter } from 'react-router-dom';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';
import ArticleCard from 'Components/atoms/ArticleCard';
import { getArticle } from 'Actions/article';
import './style.scss';

class ArticleContainer extends Component {
  constructor(props) {
    super(props);
    const { match: { params: category } } = this.props;
    this.state = {
      category
    };
  }

  componentDidMount() {
    const { getArticles, match } = this.props;
    getArticles(match.params.category);
  }

  componentWillReceiveProps(nextProps) {
    const { match } = this.props;
    if (match.params.category !== nextProps.match.params.category) {
      this.setState({
        category: nextProps.match.params.category
      });
    }
  }

  componentDidUpdate(prevProps, prevState) {
    const { getArticles } = this.props;
    const { category } = this.state;
    if (category !== prevState.category) {
      getArticles(category);
    }
  }

  render() {
    const	{	articles,	loading	}	=	this.props;

    const mappedArticleCards = articles.map((article) => {
      const {
        imageUrl, slug, title, description, createdAt, User: { firstname, lastname }
      } = article;
      return (
        <ArticleCard
          key={slug}
          img={imageUrl}
          title={title}
          description={description}
          name={`${firstname} ${lastname}`}
          time={createdAt}
        />
      );
    });
    return (
      <div className="container article-inner-container">
        {mappedArticleCards}
        {articles.length < 1 && (loading && <div className="article-loader" />)}
        {articles.length < 1 && (
          !loading
					&& (
					<div className="article-container-card">
            <i className="fas fa-exclamation notification-icon" />
            <p>This category has no articles</p>
					</div>
					)
        )}
      </div>
    );
  }
}

const mapStateToProps = state => ({
  loading: state.article.loading,
  articles: state.article.articles
});

const mapDispatchToProps = dispatch => ({
  getArticles: category => dispatch(getArticle(category))
});

ArticleContainer.propTypes = {
  articles: PropTypes.array.isRequired,
  loading: PropTypes.bool.isRequired,
  match: PropTypes.object.isRequired,
  getArticles: PropTypes.func.isRequired
};

export default withRouter(
  connect(
    mapStateToProps,
    mapDispatchToProps
  )(ArticleContainer)
);
