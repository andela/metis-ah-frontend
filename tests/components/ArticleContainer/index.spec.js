import React from 'react';
import { BrowserRouter } from 'react-router-dom';
import { Provider } from 'react-redux';
import { mount } from 'enzyme';
import thunk from 'redux-thunk';
import configureStore from 'redux-mock-store';
import ArticleContainer from 'Components/compounds/ArticleContainer';

const initialState = {
  article: {
    articles: [
      {
        slug: 'asasa',
        imageUrl: 'rectangle1',
        title: "The Internet's problems Can't Be solved with an Algorithm",
        description: "We can't Keep blaming behavior on the robots",
        createdAt: '23-09-2018',
        User: {
          firstname: 'Tomi',
          lastname: 'Adebanjo'
        }
      },
      {
        slug: 'asasaa',
        imageUrl: 'rectangle2',
        title: "The Internet's problems Can't Be solved with an Algorithm",
        description: "We can't Keep blaming behavior on the robots",
        createdAt: '23-09-2018',
        User: {
          firstname: 'Tomi',
          lastname: 'Adebanjo'
        }
      }
    ],
    loading: false
  }
};

const middleware = [thunk];
const mockStore = configureStore(middleware);
let wrapper;
let store;
let timeFormat;

describe('article container component test', () => {
  beforeEach(() => {
    timeFormat = jest.fn();
    store = mockStore(initialState);
    wrapper = mount(
      <Provider store={store}>
        <BrowserRouter>
          <ArticleContainer timeFormat={timeFormat} />
        </BrowserRouter>
      </Provider>
    );
  });

  it('should render the article container correctly', () => {
    expect(wrapper).toMatchSnapshot();
  });
});
