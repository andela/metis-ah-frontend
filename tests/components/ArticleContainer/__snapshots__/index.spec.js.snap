// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`article container component test should render the article container correctly 1`] = `
<Provider
  store={
    Object {
      "clearActions": [Function],
      "dispatch": [Function],
      "getActions": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
    }
  }
>
  <BrowserRouter>
    <Router
      history={
        Object {
          "action": "POP",
          "block": [Function],
          "createHref": [Function],
          "go": [Function],
          "goBack": [Function],
          "goForward": [Function],
          "length": 1,
          "listen": [Function],
          "location": Object {
            "hash": "",
            "pathname": "/",
            "search": "",
            "state": undefined,
          },
          "push": [Function],
          "replace": [Function],
        }
      }
    >
      <withRouter(Connect(ArticleContainer))
        timeFormat={[MockFunction]}
      >
        <Route>
          <Connect(ArticleContainer)
            history={
              Object {
                "action": "POP",
                "block": [Function],
                "createHref": [Function],
                "go": [Function],
                "goBack": [Function],
                "goForward": [Function],
                "length": 1,
                "listen": [Function],
                "location": Object {
                  "hash": "",
                  "pathname": "/",
                  "search": "",
                  "state": undefined,
                },
                "push": [Function],
                "replace": [Function],
              }
            }
            location={
              Object {
                "hash": "",
                "pathname": "/",
                "search": "",
                "state": undefined,
              }
            }
            match={
              Object {
                "isExact": true,
                "params": Object {},
                "path": "/",
                "url": "/",
              }
            }
            timeFormat={[MockFunction]}
          >
            <ArticleContainer
              articles={
                Array [
                  Object {
                    "User": Object {
                      "firstname": "Tomi",
                      "lastname": "Adebanjo",
                    },
                    "createdAt": "23-09-2018",
                    "description": "We can't Keep blaming behavior on the robots",
                    "imageUrl": "rectangle1",
                    "slug": "asasa",
                    "title": "The Internet's problems Can't Be solved with an Algorithm",
                  },
                  Object {
                    "User": Object {
                      "firstname": "Tomi",
                      "lastname": "Adebanjo",
                    },
                    "createdAt": "23-09-2018",
                    "description": "We can't Keep blaming behavior on the robots",
                    "imageUrl": "rectangle2",
                    "slug": "asasaa",
                    "title": "The Internet's problems Can't Be solved with an Algorithm",
                  },
                ]
              }
              getArticles={[Function]}
              history={
                Object {
                  "action": "POP",
                  "block": [Function],
                  "createHref": [Function],
                  "go": [Function],
                  "goBack": [Function],
                  "goForward": [Function],
                  "length": 1,
                  "listen": [Function],
                  "location": Object {
                    "hash": "",
                    "pathname": "/",
                    "search": "",
                    "state": undefined,
                  },
                  "push": [Function],
                  "replace": [Function],
                }
              }
              loading={false}
              location={
                Object {
                  "hash": "",
                  "pathname": "/",
                  "search": "",
                  "state": undefined,
                }
              }
              match={
                Object {
                  "isExact": true,
                  "params": Object {},
                  "path": "/",
                  "url": "/",
                }
              }
              timeFormat={[MockFunction]}
            >
              <div
                className="container article-inner-container"
              >
                <ArticleCard
                  description="We can't Keep blaming behavior on the robots"
                  img="rectangle1"
                  key="asasa"
                  name="Tomi Adebanjo"
                  time="23-09-2018"
                  title="The Internet's problems Can't Be solved with an Algorithm"
                >
                  <div
                    className="articleCard-container"
                  >
                    <div
                      className="columns articleCard"
                    >
                      <div
                        className="column is-one-fifth padding__0"
                      >
                        <img
                          alt=""
                          className="articleCard__img"
                          src="rectangle1"
                        />
                      </div>
                      <div
                        className="column padding__0 articleCard__content"
                      >
                        <h2
                          className="articleCard__title"
                        >
                          The Internet's problems Can't Be solved with an Algorithm
                        </h2>
                        <p
                          className="articleCard__subtitle"
                        >
                          We can't Keep blaming behavior on the robots
                        </p>
                        <p
                          className="articleCard__name"
                        >
                          Tomi Adebanjo
                        </p>
                        <p
                          className="articleCard__time"
                        >
                          just now
                        </p>
                      </div>
                    </div>
                  </div>
                </ArticleCard>
                <ArticleCard
                  description="We can't Keep blaming behavior on the robots"
                  img="rectangle2"
                  key="asasaa"
                  name="Tomi Adebanjo"
                  time="23-09-2018"
                  title="The Internet's problems Can't Be solved with an Algorithm"
                >
                  <div
                    className="articleCard-container"
                  >
                    <div
                      className="columns articleCard"
                    >
                      <div
                        className="column is-one-fifth padding__0"
                      >
                        <img
                          alt=""
                          className="articleCard__img"
                          src="rectangle2"
                        />
                      </div>
                      <div
                        className="column padding__0 articleCard__content"
                      >
                        <h2
                          className="articleCard__title"
                        >
                          The Internet's problems Can't Be solved with an Algorithm
                        </h2>
                        <p
                          className="articleCard__subtitle"
                        >
                          We can't Keep blaming behavior on the robots
                        </p>
                        <p
                          className="articleCard__name"
                        >
                          Tomi Adebanjo
                        </p>
                        <p
                          className="articleCard__time"
                        >
                          just now
                        </p>
                      </div>
                    </div>
                  </div>
                </ArticleCard>
              </div>
            </ArticleContainer>
          </Connect(ArticleContainer)>
        </Route>
      </withRouter(Connect(ArticleContainer))>
    </Router>
  </BrowserRouter>
</Provider>
`;
